utils/src/main/scala/io/udash/utils/FilteringUtils.scala

```diff
@@ -1,31 +1,15 @@
 package io.udash.utils
 
-import scala.annotation.tailrec
+import com.avsystem.commons._
 
 object FilteringUtils {
   /** Finds the longest lists prefix with equal elements. */
-  def findEqPrefix[T](newPath: List[T], previousPath: List[T]): List[T] = {
-    @tailrec
-    def _findEqPrefix(newPath: List[T], previousPath: List[T], subPath: List[T]): List[T] = {
-      (newPath, previousPath) match {
-        case (head1 :: tail1, head2 :: tail2) if head1 == head2 => _findEqPrefix(tail1, tail2, subPath :+ head1)
-        case _ => subPath
-      }
-    }
-
-    _findEqPrefix(newPath, previousPath, Nil)
-  }
+  def findEqPrefix[T](newPath: Iterator[T], previousPath: Iterator[T]): Iterator[T] =
+    newPath.zip(previousPath).takeWhile { case (h1, h2) => h1 == h2 }.map(_._1)
 
   /** Finds @newPath suffix which is different than in @previousPath. */
-  def findDiffSuffix[T](newPath: List[T], previousPath: List[T]): List[T] = {
-    @tailrec
-    def _findDiffSuffix(newPath: List[T], previousPath: List[T], subPath: List[T]): List[T] = (newPath, previousPath) match {
-      case (head1 :: tail1, head2 :: tail2) if subPath == Nil && head1 == head2 => _findDiffSuffix(tail1, tail2, subPath)
-      case (head1 :: tail1, _ :: tail2) => _findDiffSuffix(tail1, tail2, subPath :+ head1)
-      case (head1 :: tail1, Nil) => _findDiffSuffix(tail1, Nil, subPath :+ head1)
-      case _ => subPath
-    }
-
-    _findDiffSuffix(newPath, previousPath, Nil)
-  }
+  def findDiffSuffix[T](newPath: Iterator[T], previousPath: Iterator[T]): Iterator[T] =
+    newPath.map(_.opt).zipAll(previousPath.map(_.opt), Opt.Empty, Opt.Empty).dropWhile { case (h1, h2) =>
```

Comment:
possible corner-case (didn't check but worth thinking twice or writing a test):
`newPath` with last element `null` and `previousPath` of size `newPath.size - 1` should give different result than former implementation.

Reply:
Not supposed to be null-safe at all, the `.opt` calls are there just to lift and account for different lengths of these iterators.